File: built-in plugins dsp/mb-stereo-widener.cpp
/**
 * MB Stereo Widener
 * Category : effect
 * Type     : stereo
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Frequency-dependent stereo width enhancement
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_STEREO_WIDENER_H
#define MB_STEREO_WIDENER_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbStereoWidener : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-stereo-widener";
    static constexpr const char* PLUGIN_NAME    = "MB Stereo Widener";
    static constexpr const char* PLUGIN_TYPE    = "stereo";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float width = 1.2f;  // range [0, 3]
    float lowWidth = 0.8f;  // range [0, 2]
    float highWidth = 1.5f;  // range [0, 3]
    float mix = 1f;  // range [0, 1]
    };

    MbStereoWidener() = default;
    ~MbStereoWidener() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.width = std::clamp(params.width, 0f, 3f);
        params.lowWidth = std::clamp(params.lowWidth, 0f, 2f);
        params.highWidth = std::clamp(params.highWidth, 0f, 3f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Stereo Widener
        return input;
    }
};

#endif // MB_STEREO_WIDENER_H

File: built-in plugins dsp/mb-stereo-widener.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbStereoWidenerPlugin: PluginDefinition = { id: 'mb-stereo-widener', slug: 'mb-stereo-widener', name: 'MB Stereo Widener', category: 'effect', type: 'stereo' as any, version: '1.0.0', description: 'Frequency-dependent stereo width enhancement', author: 'Max Booster', grade: 'A', parameters: [{ id: 'width', name: 'Width', type: 'float', defaultValue: 1.2, minValue: 0, maxValue: 3, automatable: true }, { id: 'lowWidth', name: 'Low Width', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 2, automatable: true }, { id: 'highWidth', name: 'High Width', type: 'float', defaultValue: 1.5, minValue: 0, maxValue: 3, automatable: true }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 1, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { width: 1.2, lowWidth: 0.8, highWidth: 1.5, mix: 1 } };

export default MbStereoWidenerPlugin;
